subsys/fs/fs.c

```diff
@@ -0,0 +1,529 @@
+		}
+	}
+
+	return rc;
+}
+
+int fs_seek(fs_file_t *zfp, off_t offset, int whence)
+{
+	int rc = -EINVAL;
+
+	if (zfp->fops->lseek != NULL) {
+		rc = zfp->fops->lseek(zfp, offset, whence);
+		if (rc < 0) {
+			SYS_LOG_ERR("file seek error (%d)\n", rc);
+		}
+	}
+
+	return rc;
+}
+
+off_t fs_tell(fs_file_t *zfp)
+{
+	int rc = -EINVAL;
+
+	if (zfp->fops->tell != NULL) {
+		rc = zfp->fops->tell(zfp);
+		if (rc < 0) {
+			SYS_LOG_ERR("file tell error (%d)\n", rc);
+		}
+	}
+
+	return rc;
+}
+
+int fs_truncate(fs_file_t *zfp, off_t length)
+{
+	int rc = -EINVAL;
+
+	if (zfp->fops->truncate != NULL) {
+		rc = zfp->fops->truncate(zfp, length);
+		if (rc < 0) {
+			SYS_LOG_ERR("file truncate error (%d)\n", rc);
+		}
+	}
+
+	return rc;
+}
+
+int fs_sync(fs_file_t *zfp)
+{
+	int rc = -EINVAL;
+
+	if (zfp->fops->sync != NULL) {
+		rc = zfp->fops->sync(zfp);
+		if (rc < 0) {
+			SYS_LOG_ERR("file sync error (%d)\n", rc);
+		}
+	}
+
+	return rc;
+}
+
+/* Directory operations */
+int fs_opendir(fs_dir_t *zdp, const char *abs_path)
+{
+	fs_mount_t *mp;
+	const char *fs_path;
+	int rc, match_len;
+
+	if ((strlen(abs_path) <= 1) || (abs_path[0] != '/')) {
+		SYS_LOG_ERR("invalid file name!!\n");
+		return -EINVAL;
+	}
+
+	rc = get_mnt_point(&mp, abs_path, &match_len);
+	if (rc < 0) {
+		SYS_LOG_ERR("%s:mount point not found!!\n", __func__);
+		return rc;
+	}
+	fs_path = &abs_path[match_len];
+
+	zdp->mp = mp;
+	zdp->dops = mp->fs->dops;
+
+	if (zdp->dops->opendir != NULL) {
+		rc = zdp->dops->opendir(zdp, fs_path);
+		if (rc < 0) {
+			SYS_LOG_ERR("directory open error (%d)\n", rc);
+		}
+	}
+
+	return rc;
+}
+
+int fs_readdir(fs_dir_t *zdp, struct fs_dirent *entry)
+{
+	int rc = -EINVAL;
+
+	if (zdp->dops->readdir != NULL) {
+		rc = zdp->dops->readdir(zdp, entry);
+		if (rc < 0) {
+			SYS_LOG_ERR("directory read error (%d)\n", rc);
+		}
+	}
+	return rc;
+}
+
+int fs_closedir(fs_dir_t *zdp)
+{
+	int rc = -EINVAL;
+
+	if (zdp->dops->closedir != NULL) {
+		rc = zdp->dops->closedir(zdp);
+		if (rc < 0) {
+			SYS_LOG_ERR("directory close error (%d)\n", rc);
+			return rc;
+		}
+	}
+
+	zdp->mp = NULL;
+	zdp->dops = NULL;
+	return rc;
+}
+
+/* Filesystem operations */
+int fs_mkdir(const char *abs_path)
+{
+	fs_mount_t *mp;
+	const char *fs_path;
+	int rc, match_len;
+
+	if ((strlen(abs_path) <= 1) || (abs_path[0] != '/')) {
+		SYS_LOG_ERR("invalid file name!!\n");
+		return -EINVAL;
+	}
+
+	rc = get_mnt_point(&mp, abs_path, &match_len);
+	if (rc < 0) {
+		SYS_LOG_ERR("%s:mount point not found!!\n", __func__);
+		return rc;
+	}
+	fs_path = &abs_path[match_len];
+
+	if (mp->fs->fs_ops->mkdir != NULL) {
+		rc = mp->fs->fs_ops->mkdir(mp, fs_path);
+		if (rc < 0) {
+			SYS_LOG_ERR("failed to create directory (%d)\n", rc);
+		}
+	}
+
+	return rc;
+}
+
+int fs_unlink(const char *abs_path)
+{
+	fs_mount_t *mp;
+	const char *fs_path;
+	int rc, match_len;
+
+	if ((strlen(abs_path) <= 1) || (abs_path[0] != '/')) {
+		SYS_LOG_ERR("invalid file name!!\n");
```

Comment:
SYS_LOG_ERR will append the newline if not explicitly disabled.
I suppose "\n" is unnecessary.